### ISSUE TITLE
about entropy calculation.

### ISSUE THREAD

**Issue by DISAPPEARED13** [Haochen-Wang409/U2PL#36]
Hello, Wang. Long time no see. I have to bother you again!
When calculate contrast loss, we need to get the entropy map and filt it with alpha_t, with this code below:
```                with torch.no_grad():
                    prob = torch.softmax(pred_u_large_teacher, dim=1)
                    entropy = -torch.sum(prob * torch.log(prob + 1e-10), dim=1)

                    low_thresh = np.percentile(entropy[label_u_aug != 255].cpu().numpy().flatten(), alpha_t)
                    low_entropy_mask = (entropy.le(low_thresh).float() * (label_u_aug != 255).bool())

                    high_thresh = np.percentile(entropy[label_u_aug != 255].cpu().numpy().flatten(),100 - alpha_t)
                    high_entropy_mask = (entropy.ge(high_thresh).float() * (label_u_aug != 255).bool())

                    low_mask_all = torch.cat(((label_l.unsqueeze(1) != 255).float(), low_entropy_mask.unsqueeze(1),))
                    high_mask_all = torch.cat(((label_l.unsqueeze(1) != 255).float(), high_entropy_mask.unsqueeze(1),))
```

But I save the output of ```low_mask_all``` and ```high_mask_all```, I found something strange. 

For example, as we know that the ideal entropy map should be tend to be 0 at every class, but at least we want that **low-entropy** representations the target classes we want.

And the result ``` low_mask_all``` shows nothing (all zero map), but ```high_mask_all```shows the frame of the segmentation target! why? Is it the converse of the real result, 

like that **low_mask** show the frame of the objects, 
and **high_mask** maybe looks like pepper-salt noise or sth.

Best,

**Comment by Haochen-Wang409** [comment]
Sorry, it seems that I fail to get your problem... Could you provide a further explaination?
Did you mean that **low_mask_all** equals to 0 at each pixel?

**Comment by DISAPPEARED13** [comment]
Sorry I missed some variables, I mean **low_valid_pixel** comes from ** low_mask_all** .

I means that **low_entropy_mask** is very important and when it multiply with **label_u**, we can found it easier to get the object outline with saving **low_entropy_mask** and **high_entropy_mask**, 

but the result is very unexpected and unreasonable, while **low_valid_pixel** mostly all equals to zero as each pixel, and **high_valid_pixel** show the outline of the objects. And our problem is to made the objects to be segmented be **low entropy**, right? so they should be "True" in **low_valid_pixel**.


And with the figure of **low_entropy_mask** I saw that the object pixels are 0(False) while **high_entropy_mask** is 1(True).

**Comment by Haochen-Wang409** [comment]
You could print ```alpha_t``` while training.

In fact, it is true that **low_mask_all** and **high_mask_all** indicate **reliable** and **unreliable** pixels respectively.
However, the **low_mask_all** is defined as **the top alpha_t% RELIABLE pixels**, which means if alpha_t equals to 10, 10 percent of **low_mask_all** should be 1 and the remaining should be 0. 
And the **high_mask_all** is defined as **the top alpha_t% UNRELIABLE pixels**, which means if alpha_t equals to 10, 10 percent of **high_mask_all** should be 1 and the remaining should be 0. 

I'm not sure about the meaning ``low_valid_pixel **mostly** all equals to 0 as each pixel''. Did you mean that only 10 percent of low_valid_pixel equals to 1 (when alpha_t equals to 10)? If so, this result is as expected.
Or, could you provide the figure of masks you have saved and the corresponding input image?

Usually, the center of an object would be reliable and the borders of objects might be unreliable.
Take Fig. 2a in our paper as an instance, the blue pixels are tend to have low entropy scores, while the red ones are unreliable pixels.

**Comment by DISAPPEARED13** [comment]
Yes, thanks a lot.
with **low_mask_all** and **high_mask_all** selected top **ALPHA_T**% pixels at each time, and it can be counted at varaible ** low_entropy_mask**, but when run 
```low_mask_all = torch.cat(((label_l!=255).float().cuda(),low_entropy_mask), dim=0)```
```low_valid_pixel = torch.cat((label_l, label_u), dim=0) * low_mask_all```

low_valid_pixel mostly got 0 at foreground classes, but in background classes can saw the borders(which fit fig.2a's describtion). And label_u means pseudo-label from teacher model's inference, and I make sure that it's almost correct segmented.

But when run **low_valid_pixel**, there is only 3.3%(I set alpha_0 as 20%) pixels got **True** boolean while others got zero, and I don't think it reasonable, and this is why my contrast loss all goes to 0.0 and do nothing with model.

**Comment by DISAPPEARED13** [comment]
and **low_valid_pixel** is the metric shape as [Batch_size, num_classes+1, H, W], right? and each shape[1] means the this is the low_valid prediction for each classes. But it only shows at the first num_classes which means background.

**Comment by DISAPPEARED13** [comment]
It seems that I fixed this problem, when run 
```    low_valid_pixel = torch.cat((label_l, label_u), dim=0) * low_mask_all```
```    high_valid_pixel = torch.cat((label_l, label_u), dim=0) * high_mask_all```
and put that mask on ```label_u```, we should get the image looks like fig.2b as paper said. And my work is not to work on RGB iamge but grayscale image. So I have to change the ignore index from 255 to 0(background), after that I think I got the correct map looks like fig2.b, that is low-entropy pixels inside the object while high-entropy pixels are the border of objects.
Thanks for your inspiring suggestion!
